File: worker/src/combiner.ts
import type { Transaction, StatementMetadata } from "./models"
import { extractPdf } from "./extraction"
import { computeMetrics } from "./parser"
import { score } from "./scoring"

export async function combineAndScore(pdfBuffers: ArrayBuffer[]): Promise<Record<string, unknown>> {
  const allTxns: Transaction[] = []
  const seen = new Set<string>()
  let firstMeta: StatementMetadata = {}

  for (const buf of pdfBuffers) {
    const result = await extractPdf(buf)
    if (!firstMeta.accountHolder) firstMeta = result.metadata

    for (const t of result.transactions) {
      const lower = t.description.toLowerCase()
      if (lower.includes("beginning balance") || lower.includes("ending balance")) continue
      const key = `${t.date}:${t.description.slice(0, 40)}`
      if (!seen.has(key)) {
        seen.add(key)
        allTxns.push(t)
      }
    }
  }

  allTxns.sort((a, b) => a.date.localeCompare(b.date) || a.description.localeCompare(b.description))

  const combinedMeta: StatementMetadata = {
    bankName: firstMeta.bankName,
    accountHolder: firstMeta.accountHolder,
    accountNumber: firstMeta.accountNumber,
    statementPeriodStart: allTxns[0]?.date,
    statementPeriodEnd: allTxns[allTxns.length - 1]?.date,
  }

  const metrics = computeMetrics(allTxns, combinedMeta)
  const result = score(metrics)

  return {
    success: true,
    statementsCombined: pdfBuffers.length,
    transactionCount: allTxns.length,
    dateRange: {
      from: allTxns[0]?.date,
      to: allTxns[allTxns.length - 1]?.date,
    },
    score: result.score,
    riskTier: result.riskTier,
    maxLoanAmount: result.maxLoanAmount,
    breakdown: result.breakdown,
    confidence: result.confidence,
    metrics: Object.fromEntries(
      Object.entries(metrics).map(([k, v]) => [k, String(v)])
    ),
    metadata: Object.fromEntries(
      Object.entries(combinedMeta).filter(([_, v]) => v !== undefined) as [string, string][]
    ),
  }
}

File: backend/app/core/combiner.py
from pathlib import Path

from .extraction import extract_pdf
from .models import ScoringInputs, StatementMetadata
from .parser import TransactionParser
from .scoring import ScoringEngine


def combine_and_score(pdf_paths: list[Path]) -> dict:
    """Extract transactions from multiple PDFs, merge them, and compute a combined score."""
    parser = TransactionParser()
    engine = ScoringEngine()

    all_txns = []
    seen = set()
    metadata_list = []

    for path in pdf_paths:
        result = extract_pdf(path)
        metadata_list.append(result.metadata)

        for t in result.transactions:
            desc_lower = t.description.lower()
            if any(kw in desc_lower for kw in ["beginning balance", "ending balance"]):
                continue
            key = (t.date, t.description[:40])
            if key not in seen:
                seen.add(key)
                all_txns.append(t)

    all_txns.sort(key=lambda t: (t.date, t.description))

    m = metadata_list[0] if metadata_list else StatementMetadata()
    combined_meta = StatementMetadata(
        bank_name=m.bank_name,
        account_holder=m.account_holder,
        account_number=m.account_number,
        statement_period_start=all_txns[0].date if all_txns else None,
        statement_period_end=all_txns[-1].date if all_txns else None,
    )

    metrics = parser.compute_metrics(all_txns, combined_meta)
    score = engine.score(ScoringInputs(**metrics))

    monthly = parser.compute_monthly_breakdown(all_txns)
    for m in monthly:
        if "balance" in m and m["balance"] is not None:
            m["balance"] = float(m["balance"])
        m["income"] = float(m["income"])
        m["expenses"] = float(m["expenses"])
        m["net"] = float(m["net"])

    return {
        "success": True,
        "statements_combined": len(pdf_paths),
        "transaction_count": len(all_txns),
        "date_range": {
            "from": str(all_txns[0].date) if all_txns else None,
            "to": str(all_txns[-1].date) if all_txns else None,
        },
        "score": score.score,
        "risk_tier": score.risk_tier.value,
        "max_loan_amount": float(score.max_loan_amount),
        "breakdown": score.breakdown,
        "confidence": score.confidence,
        "metrics": {k: str(v) for k, v in metrics.items()},
        "monthly": monthly,
        "metadata": combined_meta.model_dump(exclude_none=True),
    }
